Update Address decoding to fix Litecoin decode bug

cpacia opened the pull request:
Litecoin segwit addresses start with ltc1, however there is about a 1 in 40,000
chance that a legacy L address will also start with ltc1. When this happens the
address decoding fails as it make the decision to parse as segwit or parse as
legacy based on the prefix. In this case it tries and fails to parse as segwit
since the prefix is ltc.

This commit makes a small change to the parse logic to check both the prefix
and the encoding when deciding whether to parse as segwit or legacy.

Tests are updated to test for this case.

martinboehm commented:
Thank you very much for your work!

I would prefer to implement it slightly differently:

```
if chaincfg.IsBech32SegwitPrefix(prefix) {
    witnessVer, witnessProg, err := decodeSegWitAddress(addr)
    // Some non-segwit addresses (for example, Litecoin `L` addresses)
    // may also start with the segwit prefix even though they are
    // legacy addresses. In this case decodeSegWitAddress returns error
    // and we proceed to try decoding as a legacy address below.
    if err == nil {
		... the witness version etc
    }
	... in case of error we continue to the non-segwit code
}
```
This way we do not have to call decodeSegWitAddress for every address but only for potential segwit addreses. My version of the code passes your tests.

What do you think?

cpacia committed:
Update Address decoding to fix Litecoin decode bug

Litecoin segwit addresses start with ltc1, however there is about a 1 in 40,000
chance that a legacy L address will also start with ltc1. When this happens the
address decoding fails as it make the decision to parse as segwit or parse as
legacy based on the prefix. In this case it tries and fails to parse as segwit
since the prefix is ltc.

This commit makes a small change to the parse logic to check both the prefix
and the encoding when deciding whether to parse as segwit or legacy.

Tests are updated to test for this case.

cpacia commented:
@martinboehm updated

martinboehm commented:
@cpacia Thank you very much. The implementation probably works, however the test address `LTC1eqUzePT9uvpvb413Ejd6P8Cx1Ei8Di` is not detected as an address with bech32 prefix, because the code searches for `1` from the end of the string and finds the substring `LTC1eqUzePT9uvpvb413Ejd6P8Cx1`, which is not 'ltc1' (ignoring the case).

Would you have some other litecoin address with prefix `LTC1` which does not contain any other character `1`? It would be nice to add it as a test case.

martinboehm commented:
@cpacia Thank you very much for pointing out the problem. I located a `LTC1` address which does not contain any other character `1` so I can complete the test.
